### ISSUE TITLE
clean out the core zk package 

### ISSUE THREAD

**Issue by jeffbean** [go-zookeeper/zk#29]
This has already pseudo started with the repo https://github.com/go-zookeeper/recipes. 

I would suggest we write down a scheme we intend to prescribe so we know where things belong.

github.com/go-zookeeper/recipes/lock
github.com/go-zookeeper/recipes/election

We should move/remove these peripherals to other repositories or even just sub-Go packages to remove the noise of supporting recipes in the core client. 

These would be different than let's say `github.com/go-zookeeper/zk/zktestutils`, or where the FourLetter Helpers, maybe `go-zookeeper/zk/flw`. 

My hunch is that these patterns/recipes would get the most issues or questions/support, so from separating them into another repo seems smart. On the other hand that may lead to a tiny bit more overhead in redirecting or moving issue over to the other repo. On the other hand, we could have all of these be in the same repo, but sub-packages

github.com/go-zookeeper/zk/recipes/lock
github.com/go-zookeeper/zk/recipes/election

or even

github.com/go-zookeeper/zk/lock
github.com/go-zookeeper/zk/election

cc @nemith

**Comment by nemith** [comment]
I think we should group recipe subpackages together in either the same repo or in a different one.    Monorepo is life for a lot of us so the same repo probably makes more sense. 

The other thing to think of is module versioning.  Since go modules are version together do we want to version recipes along with the core library?  I don't think we do.

**Comment by jeffbean** [comment]
oh! I had not thought of versioning. we also live in a monorepo and versioning is a thing of the past. In this case, I would lean away from having to even need to have a version support matrix of core lib + pattern version to report bugs and such. 

That all being said, I tent to keep things in the same repo since that is the norm for me. So my vote is to keep this all in one repo and put these things in sub-packages, but I'm open to other setups. 

I wanted to rip out the locks :D so whatever we do I want to make sure we agree and have some consensus so we know whats going on.

**Comment by nemith** [comment]
I want to rip out locks too. I am down for for whatever direction.  Sub-package/sub directory probably is the easiest (alias the existing one for 1.x compatability, deprecate and remove it in 2.x)

One repo works for me.  Any thoughts @pmazzini ?

**Comment by pmazzini** [comment]
I would group all recipe subpackages together in the same repo.

github.com/go-zookeeper/zk/recipe

I think `recipe` is recommended over `recipes` (singular over plural):
https://rakyll.org/style-packages/

I actually started looking at this but left it half way. In order to be able to separate the locks and not have circular dependencies, more things need to be split (testutil, flw).

**Comment by jeffbean** [comment]
Yeah, i think FLW would come first, and then testutil. (why i mentioned them) this is no small change but i would like to start chipping away at it. 

I will go ahead and tease out the FLW as a diff to start to see how this ends up looking. Overall we agree it stays in this repo. And the naming can be reviewed in the respective PRs.

**Comment by jeffbean** [comment]
> I want to rip out locks too. I am down for for whatever direction. Sub-package/sub directory probably is the easiest (alias the existing one for 1.x compatability, deprecate and remove it in 2.x)

I want to acknowledge and echo the point here that all changes must be backward compatible, until we decide to cut out a 2.X. until then, yes we must make non breaking changes in anything here. 

good call out.
